language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -475,93 +661,275 @@ impl GlobalRef {
+            .unwrap()
+            .0;
+        Ok(Value::global_ref(GlobalRef::new_ref(self, field_ref)))
     }
-}
 
-impl Reference for GlobalRef {
-    fn borrow_field(&self, idx: u32) -> Option<Self> {
-        match &*self.reference.peek() {
-            Value::Struct(ref vec) => match vec.get(idx as usize) {
-                Some(field_ref) => Some(GlobalRef::new_ref(self, field_ref.shallow_clone())),
-                None => None,
-            },
-            _ => None,
-        }
+    fn copy_value(&self) -> VMResult<Value> {
+        let value = self.reference.copy_value()?;
+        Ok(value)
     }
 
-    fn read_reference(self) -> MutVal {
-        self.reference.clone()
+    fn write_value(self, value: Value) {
+        self.root.borrow_mut().mark_dirty();
+        self.reference.write_value(value);
     }
 
-    fn mutate_reference(self, v: MutVal) {
-        self.root.borrow_mut().mark_dirty();
-        self.reference.mutate_reference(v);
+    fn equals(&self, v2: &GlobalRef) -> VMResult<bool> {
+        self.reference.equals(&v2.reference)
     }
 
-    fn size(&self) -> AbstractMemorySize<GasCarrier> {
-        words_in(*REFERENCE_SIZE)
+    fn equals_ref(&self, v2: &Reference) -> VMResult<bool> {
+        self.reference.equals(&v2.0)
+    }
+
+    /// Normal code should always know what type this value has. This is made available only for
+    /// tests.
+    #[allow(non_snake_case)]
+    #[doc(hidden)]
+    fn to_type_FOR_TESTING(&self) -> Type {
+        self.reference.to_type_FOR_TESTING()
     }
 }
 
-//
-// Conversion routines for the interpreter
-//
+/// API for locals in a `Frame`.
+///
+/// Set up onf a frame includes the allocation for the vector of locals.
+/// The size of the locals is known at compile time, recorded in the binary and verified.
+/// When a method is called we read the local count and initialize the vector of locals (via
+/// `Locals::new`.
+/// We then have 4 operations (bytecodes) that operate on locals: CopyLoc, MoveLoc, StoreLoc and
+/// BorrowLoc.
+/// Each opcode has an entry point here with a function named after the opcode.
+impl Locals {
+    /// Create a `Locals` instance containing `size` locals all initialized to Invalid.
+    ///
+    /// The size of the locals is fixed and immutable. There is no API to grow or shrink the
+    /// locals vector.
+    pub fn new(size: usize) -> Self {
+        Locals(vec![ValueImpl::Invalid; size])
+    }
+
+    /// Copy the `Value` at index `idx`. Return an error if index is out of bounds.
+    pub fn copy_loc(&self, idx: usize) -> VMResult<Value> {
+        if let Some(local_ref) = self.0.get(idx) {
+            local_ref.copy_value()
+        } else {
+            let msg = format!(
+                "Index {} out of bounds for {} while indexing {}",
+                idx,
+                self.0.len(),
+                IndexKind::LocalPool
+            );
+            Err(VMStatus::new(StatusCode::INDEX_OUT_OF_BOUNDS).with_message(msg))
```

Comment:
There are several occurrences of StatusCode::INDEX_OUT_OF_BOUNDS in Locals.  These should not happen if bytecode verifier does its job properly.  Earlier, we saw INTERNAL_TYPE_ERROR being used in the situation when bytecode verifier did not do its job.  I wonder if we should just unify the two codes because the reason for them occurring is the same.  INTERNAL_TYPE_ERROR seems a fine name to me.

Reply:
we have been returning index out of bounds in those cases, and that is what I was doing.
Maybe the names are not clear enough and maybe we should review them, but overall an INTERNAL_TYPE_ERROR is when a type mismatch happens whereas a INDEX_OUT_OF_BOUNDS is when an index in bytecode went sideways...